### PR TITLE
Add LCA

### PR DESCRIPTION
**Pull request by didwhddks**
- [x] Read [Contribute to NTHU-CPP](https://nthu-cp.github.io/NTHU-CPP/others/contribution.html).
- [x] Rebase to the latest main branch.
- [x] List all the references.
- [x] Successfully build the website by mdBook with no error after the change.
- [x] Exclude any irrelevant files.
- [x] Link to the issue which is related to your change.
- [x] Review the content by myself.
- [x] Pass tests/CI.

### CODE REVIEW

**Comment by didwhddks** [comment]
@harry900831 助教好，我已經完成所有的 requested changes，想請你幫我 review 一下。另外因為我不太熟悉 git 的操作，搞砸了 commit history，所以我關閉了先前的 PR 並重新發了一次。

File: src/graph/lowest_common_ancestor.md [hunk 1]
```diff
@@ -0,0 +1,942 @@
...
+}
+
+```
+
+</details>
+
+### Tarjan's Offline Algorithm
+
+Tarjan 提出的想法是利用 Disjoint Set 來動態地合併子樹，並且用一陣列去記錄集合中深度最淺的點，這邊我們稱為集合的最高點。一開始，所有點都位於一個獨立的並查集且集合的最高點為自己本身，共 \\( N \\) 個集合。接著利用 DFS 去走訪所有點，在 DFS 的走訪過程中，針對經過的每一點 \\( u \\) 我們依序進行以下的操作：
+
```
Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:237]: 對於DFS 走訪到的點 u

File: src/graph/lowest_common_ancestor.md [hunk 2]
```diff
@@ -0,0 +1,942 @@
...
+- DFS 遞迴下去搜索點 \\( u \\) 的所有子樹節點
+- 遞迴 return 後檢查所有包含點 \\( u \\) 在內的查詢點對，如果另一點為搜索過的狀態，那麼根據 Tarjan 的想法，此時我們可以知道該查詢點對的最近共同祖先為另一點所在集合的最高點。
```
Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:239]: 子樹們的集合要與點 u 合併？

Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:239]: 若另一個點 還沒被檢查到怎麼辦？

File: src/graph/lowest_common_ancestor.md [hunk 3]
```diff
@@ -0,0 +1,942 @@
...
+  return 0;
+}
+
+```
+
+</details>
+
+### Heavy Path Decomposition
+
+此小節主要關注於如何利用樹重鏈剖分的概念來查詢點對的最近共同祖先，如果讀者對於樹重鏈的定義以及其相關操作（跳輕邊）還不太清楚的話，建議可以先去觀看相關章節。
```
Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:358]: 樹重練剖分 -> 樹練剖分

File: src/graph/lowest_common_ancestor.md [hunk 4]
```diff
@@ -0,0 +1,942 @@
...
+而對於 Case 2，點 \\( u \\) 和點 \\( v \\) 位於不同重鏈上，我們可以透過跳輕邊的方式將點 \\( u \\) 和點 \\( v \\) 調整至同一條重鏈上。假設經過跳輕邊調整之後，點 \\( u \\) 和點 \\( v \\) 分別跳至點 \\( i \\) 和點 \\( j \\)，我們可以得到以下資訊：
+
+- 點 \\( i \\) 和點 \\( j \\) 為祖孫關係，因為位於同一條重鏈上。
+- 點對 \\( (i,\ u) \\) 和點對 \\( (j,\ v) \\) 分別為祖孫關係。
+
+沒有頭緒的讀者可以看以下示意圖：
+
+<img src="image/LCA/heavy_path_decomposition_3.png" width="700" style="display:block; margin: 0 auto;"/>
+
+結合上述資訊可知，點 \\( i \\) 為點 \\( u \\) 和點 \\( v \\) 的共同祖先且點 \\( u \\) 和點 \\( v \\) 會在點 \\( i \\) 的不同子樹內，否則點 \\( u \\) 和點 \\( v \\) 在跳上來當前重鏈之前就會在同一重鏈上。因此對於 Case 2，可知點 \\( i \\) 會是點對 \\( (u,\ v) \\) 的最近共同祖先。
```
Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:379]: 透過不斷跳輕邊直到調整到同一重鏈上的正確性說明不夠完整

還是你覺得這邊不需要好好證明，這應該是樹練剖分章節該證明的？

File: src/graph/lowest_common_ancestor.md [hunk 5]
```diff
@@ -0,0 +1,942 @@
...
+
+<img src="image/LCA/RMQ_2.png" width="700" style="display:block; margin: 0 auto;"/>
+
+怎麼得出這個結論的呢？我們可以發現在歐拉路徑上從點 \\( u \\) 到點 \\( v \\) 的任意一條路徑，必定會包含點 \\( u \\) 到點 \\( v \\) 的最短路徑。根據稍早提及的最近共同祖先的特性，我們可以知道 \\( LCA(u,\ v) \\) 會位於最短路徑上且為該路徑上深度最淺的點。而除了經過最短路徑之外，我們可能會多經過有在最短路徑上的點的子樹，但由於多經過的這些點深度都比 \\( LCA(u,\ v) \\) 深。因此針對從點 \\( u \\) 到點 \\( v \\) 的任意一條路徑，\\( LCA(u,\ v) \\) 都會是路徑上深度最淺的點，所以在查詢時我們可以選擇任意一條路徑。但為了方便起見，我們可以在不影響結果的情況下針對所有查詢都選擇從第一次出現的點 \\( u \\) 到第一次出現點 \\( v \\) 的路徑。
+
+因此，透過歐拉路徑我們可以將 LCA 問題轉化成找區間最小值的問題。一般來說，我們會另開一陣列來記錄每個點在歐拉路徑上第一次出現時所對應的 index，方便我們在 \\( O(1) \\) 時間內找到對應的區間。
+
+<img src="image/LCA/RMQ_3.png" width="700" style="display:block; margin: 0 auto;"/>
+
+有了上述預處理好的陣列後，剩下的工作就會是處理 RMQ 問題。對於 RMQ 問題，有許多經典的解法像是線段樹、序列分塊或是 Sparse Table，這邊我們不會細講這些經典解法的概念和實作細節，有興趣的讀者可以去觀看相關的章節。而此小節我們主要關注於如何利用高度分塊來達到 \\( O(N) \\) 預處理、每次查詢 \\( O(1) \\) 的做法。
```
Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:508]: ...如何利用對歐拉路徑的陣列分塊....

File: src/graph/lowest_common_ancestor.md [hunk 6]
```diff
@@ -0,0 +1,942 @@
...
+
+#### O(n) -- O(1) LCA
+
+觀察 Depth 陣列我們可以發現，陣列中相鄰的值會剛好差 1。原因是因為在建立歐拉路徑時，我們每次只會往上或是往下走一步，所以在歐拉路徑陣列中相鄰的點，點深度會剛好差 1。基於此特性，我們可以用來優化區間最小值的查詢。
```
Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:512]: 觀察歐拉路進的陣列的每個元素的深度

File: src/graph/lowest_common_ancestor.md [hunk 7]
```diff
@@ -0,0 +1,942 @@
...
+對於搜索塊之間的最小值，相較於直接遍歷過 \\( l \\) 和 \\( r \\) 中間所有塊，我們可以在 \\(O\left(\frac{N}{K}\log\frac{N}{K}\right) \\) 的時間內先對 Block 陣列建 Sparse Table。基於 RMQ 的特性，之後的查詢可以在 \\( O(1) \\) 時間內辦到 。而對於在塊內搜索最小值，暴力做的話會需要 \\( O(K) \\) 的時間，有沒有更快的作法？
+
+還記得在此小節一開始我們提及的歐拉路徑的特性嗎？藉由該特性，我們其實可以透過像是建立差分數組的方式來將每一塊 block 轉換成大小為 \\( K-1 \\) 的塊，使得新塊內元素只會有 \\( +1 \\) 和 \\( -1 \\) 兩種可能性。舉例來說，假設存在一塊 block A 其內容為 \\( \left[0,1,2,1,0\right] \\)，我們可以將其轉換成內容為 \\( \left[1,1,-1,-1\right] \\) 的新塊。而一般來說我們會將 \\( -1 \\) 轉成 \\( 0 \\)，用 bitmask 的方式來方便紀錄以及識別每一種新塊。
+
+透過此轉換，我們可以發現新塊最多只會有 \\( 2^{K-1} \\) 種。而轉換後有著相同內容的塊，塊內所有區間組合的最小值在塊內所對應的 index offset 都會是相同的，因為新塊內容其實間接說明了原塊內的深度分布。舉例來說，假設存在一塊 block A 其內容為 \\( \left[0,1,0\right] \\)，一塊 block B 其內容為 \\( \left[2,3,2\right] \\)。在經過轉換之後，這兩塊會對應到相同的 bitmask，而針對這兩塊內的所有區間組合我們可以發現
+
+- 區間 \\( [1,\ 1] \\)、\\( [2,\ 2] \\) 和 \\( [3,\ 3] \\) 的最小值均分別對應到 index \\( 1 \\)、\\( 2 \\) 和 \\( 3 \\)
+- 區間 \\( [1,\ 2] \\) 的最小值均對應到 index \\( 1 \\)
+- 區間 \\( [2,\ 3] \\) 的最小值均對應到 index \\( 3 \\)
+- 區間 \\( [1,\ 3] \\) 的最小值均對應到 index \\( 1 \\)
```
Comment (harry900831) [inline src/graph/lowest_common_ancestor.md:529]: 這一個舉例看不太懂

**Review by harry900831 — CHANGES_REQUESTED** [review]
I'm thinking about whether wrapping the implementation to a struct object is a good idea, such as: https://github.com/OmeletWithoutEgg/ckiseki/blob/master/codes/Graph/BCC.cpp.
Therefore others may easily copy it to their codebook and having a better readability compare to reading lambda code.
 
 I'm thinking about the following two interface:
 
 1. Pass the graph to construct the lca object.
```cpp
struct LCA {
...
  LCA(vector<vector<int>>& g, int n);
  int lca(int u, int v);
  vector<vector<int>> ancestor;
};
```
2. store the graph inside lca object
```cpp
struct LCA {
...
  LCA(int n);
  int lca(int u, int v);
  void add_edge(int u, int v);
  vector<vector<int>> ancestor;
  vector<vector<int>> g;
};
```

File: src/graph/lowest_common_ancestor.md (hunk 4 above)
Comment (didwhddks) [reply src/graph/lowest_common_ancestor.md]: 我是想說關於樹鏈剖分相關操作的正確性以及時間複雜度都會在相關章節被證明，所以這邊就沒有過多著墨。

Comment (harry900831) [reply src/graph/lowest_common_ancestor.md]: Got it

File: src/graph/lca.md [hunk 8]
```diff
@@ -0,0 +1,1016 @@
...
+  void add_edge(int u, int v) {
+    adj[u].emplace_back(v);
+    adj[v].emplace_back(u);
+  }
+  void dfs(int u = 0, int p = -1) {
+    tin[u] = ++time;
+    ancestor[u][0] = p;
+    for (int i = 1; i <= logN; ++i)
+      ancestor[u][i] =
+          ~ancestor[u][i - 1] ? ancestor[ancestor[u][i - 1]][i - 1] : -1;
```
Comment (harry900831) [inline src/graph/lca.md:198]: If you store the vertex by 1-based, and 0 as -1 over here, then you don't have to do this kind of ?:.

I would prefer to refactor to 1-based.

File: src/graph/lca.md [hunk 9]
```diff
@@ -0,0 +1,1016 @@
...
+  }
+  void dfs(int u = 0, int p = -1) {
+    tin[u] = ++time;
+    ancestor[u][0] = p;
+    for (int i = 1; i <= logN; ++i) {
+      ancestor[u][i] =
+          ~ancestor[u][i - 1] ? ancestor[ancestor[u][i - 1]][i - 1] : -1;
+      mx[u][i] = ~ancestor[u][i - 1]
+                     ? max(mx[u][i - 1], mx[ancestor[u][i - 1]][i - 1])
+                     : mx[u][i - 1];
```
Comment (harry900831) [inline src/graph/lca.md:899]: Ditto

File: src/graph/lca.md [hunk 10]
```diff
@@ -0,0 +1,1016 @@
...
+  int time;
+  vector<int> parent, tin, tout;
+  vector<vector<int>> adj;
+  LCA(int N) {
+    parent.resize(N, -1);
+    tin.resize(N);
+    tout.resize(N);
+    adj.resize(N);
+    time = 0;
+  }
```
[the remark this answers is not in the document]
Comment (harry900831) [reply src/graph/lca.md:77]: to 1-based index

File: src/graph/lca.md [hunk 11]
```diff
@@ -0,0 +1,1016 @@
...
+- 節點為 0 - indexed 且根的所有祖先節點均為 \\( -1 \\)。
+- `tin[i]` 代表在 DFS 的過程中進入點 \\( i \\) 的時間。
+- `tout[i]` 代表在 DFS 的過程中離開點 \\( i \\) 的時間
+- `ancestor[i][j]` 代表第 \\( i \\) 個節點的第 \\( 2^j \\) 個祖先。
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct LCA {
```
[the remark this answers is not in the document]
Comment (harry900831) [reply src/graph/lca.md:177]: If you store the vertex by 1-based, and 0 as -1 over here, then you don't have to do this kind of ?:.

I would prefer to refactor to 1-based.

**Review by harry900831 — APPROVED** [review]

**Pull request merged**